### PR TITLE
Enhancement: Improve support and tests for long identifiers in the database (without enabling)

### PR DESCRIPTION
**Pull request by innerdvations**
### What does it do?

- removes the review workflow workaround that was previously used
- fixes issue where base table name for content types was being snake_case where it previously wasn't in Strapi 4
- enforces an explicit maxLength passed to all the naming functions
- improves testing of createMetadata and naming functions

Still remaining TODO: write a migration and turn the feature on by default

### Why is it needed?

Tracking issue: https://github.com/strapi/strapi/issues/13117

### How to test it?

Test that it works:
- Starting with this PR, enable the feature by changing MAX_DB_IDENTIFIER_LENGTH to 55
- Everything in Strapi should work fine
- Create content types, attributes, and every type of relation with names of 60 characters
- There should be no index, column, or table in the database with a name longer than 55 characters

Test that it still hasn't changed anything (feature is not on):
- create a v4 project
- switch to this PR (do not update the identifier length constant)
- everything should work exactly as before with no data loss

### Related issue(s)/PR(s)

Tracking issue: https://github.com/strapi/strapi/issues/13117

Previous PRs to lay the foundation:
https://github.com/strapi/strapi/pull/19608
https://github.com/strapi/strapi/pull/19596
https://github.com/strapi/strapi/pull/19567
https://github.com/strapi/strapi/pull/19561
https://github.com/strapi/strapi/pull/19560
https://github.com/strapi/strapi/pull/19460

### CODE REVIEW

**Comment by derrickmehaffy** [comment]
Haven't read through the code yet but how do we intend to migrate v4 project tables to this without data loss?

**Comment by innerdvations** [comment]
> Haven't read through the code yet but how do we intend to migrate v4 project tables to this without data loss?

We still have access to all the original full length names, so the migration will compare the map of shortened names to long names, and rename any that have changed.

File: packages/core/database/src/utils/identifiers/index.ts [hunk 1]
```diff
@@ -70,72 +69,81 @@ export const getMorphTableName = (
...
 
 /**
  * INDEXES
+ * Note that these methods are generally used to reference full table names + attribute(s), which
+ * may already be shortened strings rather than individual parts.
+ * That is fine and expected to compress the previously incompressible parts of those strings,
+ * because in these cases the relevant information is the table name and we can't really do
+ * any better; shortening the individual parts again might make it even more confusing.
+ *
+ * For example, the fk for mytable_myattr4567d_localizations would become
```
Comment (innerdvations) [inline packages/core/database/src/utils/identifiers/index.ts:118]:
```suggestion
 * So for example, the fk for the table `mytable_myattr4567d_localizations` will become
```

File: packages/core/database/src/index.ts [hunk 2]
```diff
@@ -10,14 +10,18 @@ import { createConnection } from './connection';
...
 import { Callback, transactionCtx, TransactionObject } from './transaction-context';
 import { validateDatabase } from './validations';
-import type { Model } from './types';
+import type { MetadataOptions, Model } from './types';
 import * as identifiers from './utils/identifiers';
 
 export { isKnexQuery } from './utils/knex';
 
+// The max length for a database identifier
+export const MAX_DB_IDENTIFIER_LENGTH = 0;
```
Comment (Convly) [inline packages/core/database/src/index.ts:19]: Could we add a comment about why we're defaulting that to 0?

File: packages/core/database/src/validations/index.ts [hunk 3]
```diff
@@ -1,10 +1,11 @@
 import { validateRelations } from './relations';
 
 import type { Database } from '..';
+import { MetadataOptions } from '../types';
```
Comment (Convly) [inline packages/core/database/src/validations/index.ts:4]:
```suggestion
import type { MetadataOptions } from '../types';
```

File: packages/core/database/src/validations/relations/index.ts [hunk 4]
```diff
@@ -1,11 +1,12 @@
 import { validateBidirectionalRelations } from './bidirectional';
 
 import type { Database } from '../..';
+import { MetadataOptions } from '../../types';
```
Comment (Convly) [inline packages/core/database/src/validations/relations/index.ts:4]:
```suggestion
import type { MetadataOptions } from '../../types';
```

File: packages/core/database/src/schema/schema.ts [hunk 5]
```diff
@@ -19,11 +19,11 @@ import type { Attribute } from '../types';
  * any other potential manipulation to it
  * */
 
-const createColumn = (name: string, attribute: Attribute): Column => {
+const createColumn = (name: string, attribute: Attribute, options: MetadataOptions): Column => {
   const { type, args = [], ...opts } = getColumnType(attribute);
 
   return {
-    name,
+    name: identifiers.getColumnName(name, options), // TODO: is this right?
```
Comment (Convly) [inline packages/core/database/src/schema/schema.ts:26]: Do we want to keep this todo? Is this actually right?

File: packages/core/database/src/schema/index.ts [hunk 6]
```diff
@@ -6,6 +6,7 @@ import createSchemaStorage from './storage';
 import { metadataToSchema } from './schema';
 
 import type { Database } from '..';
+import { MetadataOptions } from '../types';
```
Comment (Convly) [inline packages/core/database/src/schema/index.ts:9]:
```suggestion
import type { MetadataOptions } from '../types';
```

**Review by christiancp100 — APPROVED** [review]

**Comment by innerdvations** [comment]
I've found a few bugs in here while working on the migration itself; will fix them first

**Comment by innerdvations** [comment]
Due to the errors in this PR, I'm closing it in favor of https://github.com/strapi/strapi/pull/19732 which includes the fixes and the migration, since they were both expected to be merged together anyway.

**Pull request closed without merging**